dashboard/app/models/user_level.rb

```diff
@@ -138,15 +130,31 @@ def before_unsubmit
     end
   end
 
-  def has_autolocked?(stage)
-    return false unless stage.lockable?
-    unlocked_at && unlocked_at < AUTOLOCK_PERIOD.ago
+  # `locked` is a virtual attribute because it relies on `unlocked_at` to
+  # automatically return `true` after `AUTOLOCK_PERIOD`, so the following
+  # are its getter and setter
+  def locked
+    unlocked_at.nil? || unlocked_at < AUTOLOCK_PERIOD.ago
   end
 
-  def locked?(stage)
+  def locked=(is_locked)
+    self.unlocked_at = is_locked ? nil : Time.now
+  end
+
+  # this is the "locked" value we return to the client. if the lesson isn't
+  # lockable or we are an authorized teacher, we always return `false`.
+  def show_as_locked?(stage)
     return false unless stage.lockable?
     return false if user.authorized_teacher?
-    submitted? && !readonly_answers? || has_autolocked?(stage)
+    locked
+  end
+
+  # `readonly` and `locked` are mutually exclusive on the client, so we use
+  # this helper to override the value of `readonly_answers` when we're supposed
+  # to show as locked.
+  def show_as_readonly?(stage, is_locked)
+    self.locked = is_locked
```

Comment:
It seems unexpected that calling `show_as_readonly?` would update the locked value?  Actually, why do we need to pass this value in -- shouldn't the object know if it's locked or not?

Reply:
I was just playing with this to see if it would help fix an error we were seeing locally, but now it appears it might be a dev environment issue and not a bug with this code. When we are calling this method in lesson.rb, the locked value hasn't been updated yet, so I worry this method isn't working the way we want it to.